### PR TITLE
DOC: DataFrame.reindex improve summary and return type

### LINKED ISSUE #33888: DOC:https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.reindex.html#pandas.DataFrame.reindex
#### Location of the documentation

https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.reindex.html#pandas.DataFrame.reindex

#### Documentation problem
Returns
DataFrame with changed index.

#### Suggested fix for documentation
fix= a NEW DataFrame instance with changed index,NOT changing old instance .
why? I thought old instance has changed index.

### PR DESCRIPTION
**Pull request by gauravchandok**
- [x] closes #33888 
- [x] tests added / passed (passes `scripts/validate_docstrings.py pandas.DataFrame.reindex`)
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry (Not Required)

### CODE REVIEW

File: pandas/core/generic.py [hunk 1]
```diff
@@ -4241,9 +4241,9 @@ def reindex(self: FrameOrSeries, *args, **kwargs) -> FrameOrSeries:
         """
         Conform %(klass)s to new index with optional filling logic.
 
-        Places NA/NaN in locations having no value in the previous index. A new object
-        is produced unless the new index is equivalent to the current one and
-        ``copy=False``.
+        Places NA/NaN in locations having no value in the previous index. A new
+        DataFrame with an additional index is generated subject to customization
```
Comment (fujiaxiang) [inline pandas/core/generic.py:4245]: No `reindex` does not generate additional index for `DataFrame` or `Series`. "A new object" actually refers to a new `DataFrame` or `Series`, not a new index. Perhaps you can try the function to see it's behavior and then decide your wordings

File: pandas/core/generic.py [hunk 2]
```diff
@@ -4286,7 +4286,7 @@ def reindex(self: FrameOrSeries, *args, **kwargs) -> FrameOrSeries:
 
         Returns
         -------
-        %(klass)s with changed index.
+        %(klass)s with 2 indices.
```
Comment (fujiaxiang) [inline pandas/core/generic.py:4289]: same as above, `reindex` does not give 2 indices

**Review by fujiaxiang — CHANGES_REQUESTED** [review]

**Comment by fujiaxiang** [comment]
While I'm reading your changes I realized there is more to this confusion than just the docstring. See #34663. The behavior of `copy` is confusing in my opinion. We probably need to address that first. 

At the moment, I don't think this PR explains its behavior accurately. Let's discuss in #34663 what the desired behavior should be first, which could be to deprecate `copy`. After that you can continue working on this PR or open a new PR to achieve that.

**Comment by dsaxton** [comment]
@gauravchandok Can you merge master to fix conflicts?

File: pandas/core/generic.py [hunk 3]
```diff
@@ -4239,11 +4239,11 @@ def sort_values(
 
     def reindex(self: FrameOrSeries, *args, **kwargs) -> FrameOrSeries:
         """
-        Conform %(klass)s to new index with optional filling logic.
+        Create an object with a new index and columns retained from %(klass)s.
```
Comment (jreback) [inline pandas/core/generic.py:4242]: this is not true, you can specify index and/or columns.

File: pandas/core/generic.py [hunk 4]
```diff
@@ -4239,11 +4239,11 @@ def sort_values(
...
 
-        Places NA/NaN in locations having no value in the previous index. A new object
-        is produced unless the new index is equivalent to the current one and
-        ``copy=False``.
+        The values are only retained for index that matches in %(klass)s else NA/Nan.
```
Comment (jreback) [inline pandas/core/generic.py:4244]: can you say what the intent is here?

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by dsaxton** [comment]
@gauravchandok Can you address comments and merge master?

**Comment by arw2019** [comment]
Closing for now. @gauravchandok ping us whenever you'd like to continue and we'll reopen!

**Pull request closed without merging**